chain/rosetta-rpc/src/adapters/mod.rs

```diff
@@ -144,79 +146,313 @@ pub(crate) async fn convert_block_to_transactions(
     use near_primitives::views::StateChangeCauseView;
     match cause {
         StateChangeCauseView::TransactionProcessing { tx_hash } => {
-            Ok(TransactionIdentifier::transaction(&tx_hash))
+            Ok((TransactionIdentifier::transaction(&tx_hash), Some(tx_hash)))
         }
         StateChangeCauseView::ActionReceiptProcessingStarted { receipt_hash }
         | StateChangeCauseView::ActionReceiptGasReward { receipt_hash }
         | StateChangeCauseView::ReceiptProcessing { receipt_hash }
         | StateChangeCauseView::PostponedReceipt { receipt_hash } => {
-            Ok(TransactionIdentifier::receipt(&receipt_hash))
+            Ok((TransactionIdentifier::receipt(&receipt_hash), None))
         }
         StateChangeCauseView::InitialState => {
-            Ok(TransactionIdentifier::block_event("block", block_hash))
+            Ok((TransactionIdentifier::block_event("block", block_hash), None))
         }
         StateChangeCauseView::ValidatorAccountsUpdate => {
-            Ok(TransactionIdentifier::block_event("block-validators-update", block_hash))
+            Ok((TransactionIdentifier::block_event("block-validators-update", block_hash), None))
         }
         StateChangeCauseView::UpdatedDelayedReceipts => {
-            Ok(TransactionIdentifier::block_event("block-delayed-receipts", block_hash))
+            Ok((TransactionIdentifier::block_event("block-delayed-receipts", block_hash), None))
         }
         StateChangeCauseView::NotWritableToDisk => {
             Err(crate::errors::ErrorKind::InternalInvariantError(
                 "State Change 'NotWritableToDisk' should never be observed".to_string(),
             ))
         }
         StateChangeCauseView::Migration => {
-            Ok(TransactionIdentifier::block_event("migration", block_hash))
+            Ok((TransactionIdentifier::block_event("migration", block_hash), None))
         }
         StateChangeCauseView::Resharding => Err(crate::errors::ErrorKind::InternalInvariantError(
             "State Change 'Resharding' should never be observed".to_string(),
         )),
     }
 }
 
-fn convert_block_changes_to_transactions(
+/// An builder object for generating lists of Rosetta transactions based on
+/// changes seen in given block.
+struct BlockChangesToTransactionsConverter<'a> {
+    map: std::collections::HashMap<String, crate::models::Transaction>,
+
+    /// Hash of the block for which list of transaction is created.  This is
+    /// used for generating Rosetta transaction identifiers for block events.
+    block_hash: &'a near_primitives::hash::CryptoHash,
+
+    /// Factory for generating [`crate::models::Transaction`] objects.  In
+    /// production code this is always Some and generates transactions with list
+    /// of related transactions.  In test code this may be None in which case
+    /// the transactions will be generated with empty related transactions list.
+    factory: Option<TransactionsFactory<'a>>,
```

Comment:
I would avoid adding Option to an otherwise strictly-typed field just for testing purposes, but I will let you make the final call here.

Reply:
Unfortunately I cannot think of a good way to deal with this. Because this uses async functions I cannot define a trait and than have a test mock that implements that trait. And `Addr<ViewClientActor>` isn’t currently mockable.